### ISSUE TITLE
3 ways to get a lexer for a filename, with unclear differences

### ISSUE THREAD

**Issue by vanschelven** [pygments/pygments#2783]
There are 3 ways (I think?) to get a lexer for a filename. From their docstrings, the differences are not clear (to me):

* `find_lexer_class_for_filename`
* `get_lexer_for_filename`
* `guess_lexer_for_filename`

what are the differences, and what is the preferred way?

**Comment by ooglyhLL** [comment]
More time for research needed, but I just found a terrible performance bottleneck in my code: Time spent in `get_lexer_for_filename()` exceeded time spent in the lexers themselves (measured over some 2900 files scanned by the same process). Not sure about the other two candidates.
As a quick workaround I created my own dict for file extension->lexer mapping (which I would have done eventually anyway, but for other reasons).

**Comment by vanschelven** [comment]
> [..] terrible performance bottleneck in my code: Time spent in `get_lexer_for_filename()` exceeded time spent in the lexers themselves

Yes I ran into that in [bugsink](https://www.bugsink.com/) recently too; I've created a separate issue for that to avoid confused discussions.
